src/components/atoms/Timeline/hooks.ts

```diff
@@ -0,0 +1,213 @@
+import React, {
+  MouseEvent,
+  MouseEventHandler,
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+  WheelEventHandler,
+} from "react";
+
+import {
+  BORDER_WIDTH,
+  DAY_SECS,
+  EPOCH_SEC,
+  SCALE_ZOOM_INTERVAL,
+  GAP_HORIZONTAL,
+  MAX_ZOOM_RATIO,
+  PADDING_HORIZONTAL,
+  SCALE_INTERVAL,
+  MINUTES_SEC,
+  HOURS_SECS,
+  NORMAL_SCALE_WIDTH,
+  STRONG_SCALE_WIDTH,
+} from "./constants";
+import { TimeEventHandler, Range } from "./types";
+
+const convertPositionToTime = (e: MouseEvent, { start, end }: Range, gapHorizontal: number) => {
+  const curTar = e.currentTarget;
+  const width = curTar.scrollWidth - (PADDING_HORIZONTAL + BORDER_WIDTH) * 2 - gapHorizontal;
+  const rect = curTar.getBoundingClientRect();
+  const clientX = e.clientX - rect.x;
+  const scrollX = curTar.scrollLeft;
+  const posX = clientX + scrollX - (PADDING_HORIZONTAL + BORDER_WIDTH);
+  const percent = posX / width;
+  const rangeDiff = end - start;
+  const sec = rangeDiff * percent;
+  return start + sec;
+};
+
+type InteractionOption = {
+  range: Range;
+  gapHorizontal: number;
+  zoom: number;
+  setZoom: React.Dispatch<React.SetStateAction<number>>;
+  onClick?: TimeEventHandler;
+  onDrag?: TimeEventHandler;
+};
+
+const useTimelineInteraction = ({
+  range: { start, end },
+  gapHorizontal,
+  zoom,
+  setZoom,
+  onClick,
+  onDrag,
+}: InteractionOption) => {
+  const [isMouseDown, setIsMouseDown] = useState(false);
+  const handleOnMouseDown = useCallback(() => {
+    setIsMouseDown(true);
+  }, []);
+  const handleOnMouseUp = useCallback(() => {
+    setIsMouseDown(false);
+  }, []);
+  const handleOnMouseMove: MouseEventHandler = useCallback(
+    e => {
+      if (!isMouseDown || !onDrag || !e.target) {
+        return;
+      }
+
+      onDrag(convertPositionToTime(e, { start, end }, gapHorizontal));
+
+      const scrollThreshold = 30;
+      const scrollAmount = 20;
+      const clientX = e.clientX;
+      const curTar = e.currentTarget;
+      const clientWidth = curTar.clientWidth;
+
+      if (clientX > clientWidth - scrollThreshold) {
+        curTar.scroll(curTar.scrollLeft + scrollAmount, 0);
+      }
+      if (clientX < scrollThreshold) {
+        curTar.scroll(curTar.scrollLeft - scrollAmount, 0);
+      }
+    },
+    [onDrag, start, end, isMouseDown, gapHorizontal],
+  );
+
+  useEffect(() => {
+    window.addEventListener("mouseup", handleOnMouseUp, { passive: true });
+    return () => {
+      window.removeEventListener("mouseup", handleOnMouseUp);
+    };
+  }, [handleOnMouseUp]);
+
+  const handleOnClick: MouseEventHandler = useCallback(
+    e => {
+      if (!onClick) {
+        return;
+      }
```

Comment:
Keeps things more concise and doesn't really degrade readability with such a simple conditional
```suggestion
      if (!onClick) return;
```